File: backend/routes/premios.js
const express = require('express');
const router  = express.Router();
const { getDb }  = require('../config/db');
const { protect } = require('../middleware/auth');
const { ObjectId } = require('mongodb');

function broadcast(action, coleccion, doc) {
  if (global.io) {
    global.io.emit('premios-update', { action, coleccion, doc, ts: new Date().toISOString() });
  }
}

/* ── GET /api/premios/activos ── */
router.get('/activos', async (req, res) => {
  try {
    const db = getDb();
    if (!db) return res.status(503).json({ success: false, message: 'BD no disponible' });
    const items = await db.collection('premios_activos').find({}).sort({ createdAt: 1 }).toArray();
    const data = items.map(d => ({ ...d, _id: d._id.toString() }));
    res.json({ success: true, data });
  } catch (e) {
    console.error('[PREMIOS] GET activos:', e);
    res.status(500).json({ success: false, message: 'Error interno' });
  }
});

/* ── POST /api/premios/activos ── */
router.post('/activos', protect, async (req, res) => {
  try {
    const db = getDb();
    if (!db) return res.status(503).json({ success: false, message: 'BD no disponible' });

    const { tipo, titulo, descripcion, categoria, monto } = req.body || {};
    const TIPOS = ['first','second','third','special','team','bonus'];
    if (!TIPOS.includes(tipo) || !titulo || !descripcion || !categoria) {
      return res.status(400).json({ success: false, message: 'Datos incompletos' });
    }

    const doc = {
      tipo,
      titulo:      String(titulo).trim(),
      descripcion: String(descripcion).trim(),
      categoria:   String(categoria).trim(),
      monto:       Number(monto) || 0,
      creadoPor:   req.user?.username || req.user?.name || 'usuario',
      createdAt:   new Date()
    };

    const result = await db.collection('premios_activos').insertOne(doc);
    doc._id = result.insertedId.toString();

    broadcast('add', 'activos', doc);
    res.json({ success: true, data: doc });
  } catch (e) {
    console.error('[PREMIOS] POST activos:', e);
    res.status(500).json({ success: false, message: 'Error interno' });
  }
});

/* ── DELETE /api/premios/activos/:id ── */
router.delete('/activos/:id', protect, async (req, res) => {
  try {
    const db = getDb();
    if (!db) return res.status(503).json({ success: false, message: 'BD no disponible' });

    const id = req.params.id;
    let filter;
    try { filter = { _id: new ObjectId(id) }; } catch (_) { filter = { _id: id }; }

    const deleted = await db.collection('premios_activos').findOneAndDelete(filter);
    if (!deleted) return res.status(404).json({ success: false, message: 'No encontrado' });

    broadcast('delete', 'activos', { _id: id });
    res.json({ success: true });
  } catch (e) {
    console.error('[PREMIOS] DELETE activos:', e);
    res.status(500).json({ success: false, message: 'Error interno' });
  }
});

/* ── GET /api/premios/ganadores ── */
router.get('/ganadores', async (req, res) => {
  try {
    const db = getDb();
    if (!db) return res.status(503).json({ success: false, message: 'BD no disponible' });
    const items = await db.collection('premios_ganadores').find({}).sort({ createdAt: 1 }).toArray();
    const data = items.map(d => ({ ...d, _id: d._id.toString() }));
    res.json({ success: true, data });
  } catch (e) {
    console.error('[PREMIOS] GET ganadores:', e);
    res.status(500).json({ success: false, message: 'Error interno' });
  }
});

/* ── POST /api/premios/ganadores ── */
router.post('/ganadores', protect, async (req, res) => {
  try {
    const db = getDb();
    if (!db) return res.status(503).json({ success: false, message: 'BD no disponible' });

    const { tipo, nombre, iniciales, monto, categoria, fecha, status } = req.body || {};
    if (!nombre || !iniciales) {
      return res.status(400).json({ success: false, message: 'Nombre e iniciales requeridos' });
    }

    const doc = {
      tipo:       tipo || 'second',
      nombre:     String(nombre).trim(),
      iniciales:  String(iniciales).trim().toUpperCase(),
      monto:      Number(monto) || 0,
      categoria:  String(categoria || '').trim(),
      fecha:      fecha || new Date().toISOString().slice(0, 10),
      status:     status === 'pendiente' ? 'pendiente' : 'asignado',
      creadoPor:  req.user?.username || req.user?.name || 'usuario',
      createdAt:  new Date()
    };

    const result = await db.collection('premios_ganadores').insertOne(doc);
    doc._id = result.insertedId.toString();

    broadcast('add', 'ganadores', doc);
    res.json({ success: true, data: doc });
  } catch (e) {
    console.error('[PREMIOS] POST ganadores:', e);
    res.status(500).json({ success: false, message: 'Error interno' });
  }
});

/* ── DELETE /api/premios/ganadores/:id ── */
router.delete('/ganadores/:id', protect, async (req, res) => {
  try {
    const db = getDb();
    if (!db) return res.status(503).json({ success: false, message: 'BD no disponible' });

    const id = req.params.id;
    let filter;
    try { filter = { _id: new ObjectId(id) }; } catch (_) { filter = { _id: id }; }

    const deleted = await db.collection('premios_ganadores').findOneAndDelete(filter);
    if (!deleted) return res.status(404).json({ success: false, message: 'No encontrado' });

    broadcast('delete', 'ganadores', { _id: id });
    res.json({ success: true });
  } catch (e) {
    console.error('[PREMIOS] DELETE ganadores:', e);
    res.status(500).json({ success: false, message: 'Error interno' });
  }
});

module.exports = router;

File: CRM_PYTHON/routers/premios.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from database_mysql import AsyncSessionLocal
from sqlalchemy import text
from deps import current_user
from typing import Optional
from datetime import datetime, date, timezone

def _utcnow() -> datetime:
    """UTC naive (reemplazo de _utcnow() deprecado en Python 3.12+)."""
    return datetime.now(timezone.utc).replace(tzinfo=None)

router = APIRouter(prefix="/api/premios", tags=["Premios"])

TIPOS_VALIDOS = {"first", "second", "third", "special", "team", "bonus"}


def _doc(row) -> dict:
    r = dict(row)
    r["_id"] = str(r.get("id", ""))
    r["id"]  = str(r.get("id", ""))
    return r


class PremioActivo(BaseModel):
    tipo: str
    titulo: str
    descripcion: str
    categoria: str
    monto: float = 0


class Ganador(BaseModel):
    nombre: str
    iniciales: str
    tipo: Optional[str] = "second"
    monto: float = 0
    categoria: Optional[str] = ""
    fecha: Optional[str] = None
    status: Optional[str] = "asignado"


# ── ACTIVOS ────────────────────────────────────────────────────────
@router.get("/activos")
async def get_activos():
    async with AsyncSessionLocal() as s:
        r = await s.execute(text("SELECT * FROM premios_activos ORDER BY created_at ASC"))
        items = [_doc(row) for row in r.mappings().all()]
    return {"success": True, "data": items}


@router.post("/activos")
async def create_activo(body: PremioActivo, user: dict = Depends(current_user)):
    if body.tipo not in TIPOS_VALIDOS:
        raise HTTPException(400, "Tipo inválido")
    now = _utcnow()
    async with AsyncSessionLocal() as s:
        res = await s.execute(text("""
            INSERT INTO premios_activos (tipo, titulo, descripcion, categoria, monto, creado_por, created_at)
            VALUES (:tipo, :titulo, :desc, :cat, :monto, :by, :now)
        """), {
            "tipo": body.tipo, "titulo": body.titulo.strip(),
            "desc": body.descripcion.strip(), "cat": body.categoria.strip(),
            "monto": body.monto, "by": user.get("username", ""), "now": now,
        })
        new_id = res.lastrowid  # antes del commit (el pool puede cambiar de conexión)
        await s.commit()
        r = await s.execute(text("SELECT * FROM premios_activos WHERE id = :id"), {"id": new_id or 0})
        row = r.mappings().first()
    return {"success": True, "data": _doc(row)}


@router.delete("/activos/{premio_id}")
async def delete_activo(premio_id: str, user: dict = Depends(current_user)):
    try:
        pid = int(premio_id)
    except ValueError:
        raise HTTPException(404, "No encontrado")
    async with AsyncSessionLocal() as s:
        r = await s.execute(text("DELETE FROM premios_activos WHERE id = :id"), {"id": pid})
        await s.commit()
        if r.rowcount == 0:
            raise HTTPException(404, "No encontrado")
    return {"success": True}


# ── GANADORES ─────────────────────────────────────────────────────
@router.get("/ganadores")
async def get_ganadores():
    async with AsyncSessionLocal() as s:
        r = await s.execute(text("SELECT * FROM premios_ganadores ORDER BY created_at ASC"))
        items = [_doc(row) for row in r.mappings().all()]
    return {"success": True, "data": items}


@router.post("/ganadores")
async def create_ganador(body: Ganador, user: dict = Depends(current_user)):
    if not body.nombre or not body.iniciales:
        raise HTTPException(400, "Nombre e iniciales requeridos")
    now = _utcnow()
    fecha_val = body.fecha or date.today().isoformat()
    async with AsyncSessionLocal() as s:
        res = await s.execute(text("""
            INSERT INTO premios_ganadores (tipo, nombre, iniciales, monto, categoria, fecha, status, creado_por, created_at)
            VALUES (:tipo, :nombre, :iniciales, :monto, :cat, :fecha, :status, :by, :now)
        """), {
            "tipo": body.tipo, "nombre": body.nombre.strip(),
            "iniciales": body.iniciales.strip().upper(), "monto": body.monto,
            "cat": (body.categoria or "").strip(), "fecha": fecha_val,
            "status": "pendiente" if body.status == "pendiente" else "asignado",
            "by": user.get("username", ""), "now": now,
        })
        new_id = res.lastrowid  # antes del commit (el pool puede cambiar de conexión)
        await s.commit()
        r = await s.execute(text("SELECT * FROM premios_ganadores WHERE id = :id"), {"id": new_id or 0})
        row = r.mappings().first()
    return {"success": True, "data": _doc(row)}


@router.delete("/ganadores/{ganador_id}")
async def delete_ganador(ganador_id: str, user: dict = Depends(current_user)):
    try:
        gid = int(ganador_id)
    except ValueError:
        raise HTTPException(404, "No encontrado")
    async with AsyncSessionLocal() as s:
        r = await s.execute(text("DELETE FROM premios_ganadores WHERE id = :id"), {"id": gid})
        await s.commit()
        if r.rowcount == 0:
            raise HTTPException(404, "No encontrado")
    return {"success": True}
